src/cargo/ops/cargo_compile/packages.rs

```diff
@@ -50,16 +50,24 @@ impl Packages {
                 .map(|id| id.to_spec())
                 .collect(),
             Packages::OptOut(opt_out) => {
-                let (mut patterns, mut names) = opt_patterns_and_names(opt_out)?;
+                let (mut patterns, mut ids) = opt_patterns_and_ids(opt_out)?;
                 let specs = ws
                     .members()
                     .filter(|pkg| {
-                        !names.remove(pkg.name().as_str()) && !match_patterns(pkg, &mut patterns)
+                        let id = ids.iter().find(|id| id.matches(pkg.package_id())).cloned();
```

Comment:
The interesting part is that none of these needs a full pkgid sped match. Comparing member name with spec.name is quite sufficient I guess?

Reply:
That is not allowed. You cannot run a dependency

Reply:
I mean the member is `foo@0.1.0`, and users want to run `foo@0.2.0`. If we only check the name, we will pick `foo@0.1.1` and run it. But I think it is an invalid package spec, we should say `foo@0.2.0` does not exist in your workspace.

Reply:
Yeah that is a valid pkgid spec not existing in the workspace. We can go further by suggesting closet member to them, maybe.

So far as I can tell, only in [this branch](https://github.com/rust-lang/cargo/blob/b2e1d3b6235c07221dd0fcac54a7b0c754ef8b11/src/cargo/ops/cargo_compile/packages.rs#L70) the given specs fall through, which makes `cargo build -p <dependency>` works. Under other branches one can never get specs beyond workspace members.

Reply:
It seems the bug from [this branch](https://github.com/rust-lang/cargo/blob/b2e1d3b6235c07221dd0fcac54a7b0c754ef8b11/src/cargo/ops/cargo_compile/packages.rs#L128). When running `cargo run -p mypackage@0.1.0`, if we compare `mypackage@0.1.0` with `mypackage`, the error message "no `mypckage@0.1.0 in this workspace" is shown, even though the package is present.
So the fix is we use spec directly instead of only comparing the name.

Reply:
Oh I see. We still need a full pkgid spec matches, as we need to point out that pkgid spec, say `foo@0.2.0` doesn't exist.

Reply:
We call [this branch](https://github.com/rust-lang/cargo/blob/b2e1d3b6235c07221dd0fcac54a7b0c754ef8b11/src/cargo/ops/cargo_compile/packages.rs#L128) from [here](https://github.com/hi-rustin/cargo/blob/3c3fe21fd26aded05fb69124fba16ec0d47d5133/src/bin/cargo/commands/run.rs#L71).

Reply:
I understand the patch, and didn't mean it's incorrect. I was trying to find a way to avoid full spec match, but maybe not worthy as it may get things complicated.